better_p2p/handshake.go

```diff
@@ -0,0 +1,162 @@
+	me := wire.NewNetAddress(&net.TCPAddr{IP: nil, Port: 0}, wire.SFNodeNetwork) // shouldn't be mode configurable?
+
+	nAddr, _ := net.ResolveTCPAddr("tcp", p.address) // address was validate already, we can omit error
+	you := wire.NewNetAddress(nAddr, wire.SFNodeNetwork)
+
+	nonce, err := wire.RandomUint64()
+	if err != nil {
+		p.l.Warn("Handshake: failed to generate nonce, send VER with 0 nonce", slog.String(errKey, err.Error()))
+	}
+
+	const lastBlock = int32(0)
+	verMsg := wire.NewMsgVersion(me, you, nonce, lastBlock)
+
+	if p.userAgentName != nil && p.userAgentVersion != nil {
+		err = verMsg.AddUserAgent(*p.userAgentName, *p.userAgentVersion)
+		if err != nil {
+			p.l.Warn("Handshake: failed to add user agent, send VER without user agent", slog.String(errKey, err.Error()))
+		}
+	}
+
+	err = wire.WriteMessage(p.lConn, verMsg, wire.ProtocolVersion, p.network)
+	if err != nil {
+		p.l.Error("Handshake failed.",
+			slog.String("reason", "failed to write VER message"),
+			slog.String(errKey, err.Error()),
+		)
+
+		return false
+	}
+
+	p.l.Debug("Sent", slogUpperString(commandKey, verMsg.Command()))
+	return true
+}
+
+//nolint:staticcheck
+func finalizeHandshake(p *Peer) (ok bool) {
+
+	// do not block goroutine on read
+	readCtx, readCancel := context.WithCancel(p.execCtx)
+	defer readCancel()
+
+	read := make(chan readMessageResult, 1)
+	readController := make(chan struct{}, 1)
+
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-readCtx.Done():
+				return
+
+			case <-readController:
+				msg, _, err := wire.ReadMessage(p.lConn, wire.ProtocolVersion, p.network)
+				read <- readMessageResult{msg, err}
+			}
+		}
+	}(readCtx)
+
+	receivedVerAck := false
+	sentVerAck := false
+
+	// wait for ACK, and VER from node send VERACK
+handshakeLoop:
+	for {
+		// "read" next message
+		readController <- struct{}{}
+
+		select {
+		// peer was stopped
+		case <-p.execCtx.Done():
+			return false
+
+		case <-time.After(1 * time.Minute):
+			p.l.Error("Handshake failed.", slog.String("reason", "handshake timeout"))
+			return false
+
+		case result := <-read:
+			if result.err != nil {
+				p.l.Error("Handshake failed.", slog.String(errKey, result.err.Error()))
+				return false
+			}
+
+			receivedVerAck, sentVerAck = onHandshakeMsg(p, result.msg, receivedVerAck, sentVerAck)
+
+			if receivedVerAck && sentVerAck {
+				break handshakeLoop
+			}
+		}
+	}
+
+	// if we exit the handshake loop, the handshake has completed successfully
+	return true
+}
+
+//nolint:staticcheck
+func onHandshakeMsg(p *Peer, msg wire.Message, outReceivedVerAck, outSentVerAck bool) (bool, bool) {
+
+	switch msg.Command() {
+	case wire.CmdVerAck:
+		p.l.Debug("Handshake: received VERACK")
+		outReceivedVerAck = true
+
+	case wire.CmdVersion:
+		p.l.Debug("Handshake: received VER")
+		if outSentVerAck {
+			p.l.Warn("Handshake: received version message after sending verack.")
+			return outReceivedVerAck, outSentVerAck
+		}
```

Comment:
Why? Messages can arrive in an incorrect order, that should not break anything. Even if the peer wants to send us a version again, after sending a VerAck, that should still not be an error IMO, it should be idempotent. WDYT? I would remove this `if`

Reply:
It’s not an error but a warningto notify that something unexpected happened. This doesn’t break anything, but we don’t handle this message again to avoid a message loop. Handshake is a defined process